drivers/sx126x/sx126x.c

```diff
@@ -0,0 +1,283 @@
+/*
+ * Copyright (C) 2021 Inria
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     drivers_sx126x
+ * @{
+ *
+ * @file
+ * @brief       Device driver implementation for the SX1261/2 LoRa radio driver
+ *
+ * @author      Alexandre Abadie <alexandre.abadie@inria.fr>
+ *
+ * @}
+ */
+
+#include <errno.h>
+
+#include "sx126x_netdev.h"
+
+#include "net/lora.h"
+#include "periph/spi.h"
+
+#include "sx126x_driver.h"
+#include "sx126x.h"
+#include "sx126x_params.h"
+
+#define ENABLE_DEBUG 0
+#include "debug.h"
+
+#ifndef CONFIG_SX126X_PKT_TYPE_DEFAULT
+#define CONFIG_SX126X_PKT_TYPE_DEFAULT          (SX126X_PKT_TYPE_LORA)
+#endif
+
+#ifndef CONFIG_SX126X_CHANNEL_DEFAULT
+#define CONFIG_SX126X_CHANNEL_DEFAULT           (868300000UL)   /* in Hz */
+#endif
+
+#ifndef CONFIG_SX126X_TX_POWER_DEFAULT
+#define CONFIG_SX126X_TX_POWER_DEFAULT          (14U)           /* in dBm */
+#endif
+
+#ifndef CONFIG_SX126X_RAMP_TIME_DEFAULT
+#define CONFIG_SX126X_RAMP_TIME_DEFAULT         (SX126X_RAMP_10_US)
+#endif
+
+void sx126x_setup(sx126x_t *dev, const sx126x_params_t *params, uint8_t index)
+{
+    netdev_t *netdev = (netdev_t *)dev;
+
+    netdev->driver = &sx126x_driver;
+    dev->params = (sx126x_params_t *)params;
+    netdev_register(&dev->netdev, NETDEV_SX126X, index);
+}
+
+static const uint16_t _bw_khz[3] = {
+    [LORA_BW_125_KHZ] = 125,
+    [LORA_BW_250_KHZ] = 250,
+    [LORA_BW_500_KHZ] = 500,
+};
+
+static uint8_t _compute_ldro(sx126x_t *dev)
+{
+    uint32_t symbol_len = (uint32_t)(1 << dev->mod_params.sf) / _bw_khz[dev->mod_params.bw - SX126X_LORA_BW_125];
+    if (symbol_len >= 16) {
+        return 0x01;
+    }
+
+    return 0x00;
+}
+
+static void sx126x_init_default_config(sx126x_t *dev)
+{
+    /* call order is important */
+    sx126x_set_pkt_type(dev, SX126X_PKT_TYPE_LORA);
+    sx126x_set_channel(dev, CONFIG_SX126X_CHANNEL_DEFAULT);
+    sx126x_pa_cfg_params_t pa_cfg = {
+        .pa_duty_cycle = 0x02,
+        .hp_max = 0x02,
+        .device_sel = 0x00,
+        .pa_lut = 0x01
```

Comment:
Magic number? can you use defines here?

Reply:
The values come from the datasheet, in the "PA Optimal Settings" section. These are related to the TX power configuration. Looking at them more in depth, the values differ between variants. I'll try to come up with something better.